File: packages/agent/src/tools/aws-instances.ts
/**
 * EC2 instance specs + right-sizing logic.
 * Specs (vCPU, RAM) are derived from the same public AWS Bulk Pricing JSON
 * that aws-pricing.ts uses — no static catalog file, no AWS credentials needed.
 */
import axios from "axios";

export const CANDIDATE_FAMILIES_V1 = [
  "t3", "t3a", "t4g",
  "m5", "m5a", "m6i", "m6a", "m7i", "m7a",
  "c5", "c5a", "c6i", "c6a", "c7i",
  "r5", "r5a", "r6i", "r6a",
];

interface InstanceSpecs {
  vcpu: number;
  ram_gb: number;
  family: string;
}

const PRICING_BASE = "https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current";

// Cache: region -> { instanceType -> InstanceSpecs }
const specsCache = new Map<string, Record<string, InstanceSpecs>>();

// In-flight deduplication: region -> Promise<specs>
// Prevents 30 concurrent batch agents from all downloading the same 350MB JSON simultaneously.
const inFlight = new Map<string, Promise<Record<string, InstanceSpecs>>>();

async function fetchRegionSpecs(region: string): Promise<Record<string, InstanceSpecs>> {
  const url = `${PRICING_BASE}/${region}/index.json`;

  const resp = await axios.get<{
    products: Record<string, {
      attributes: Record<string, string>;
    }>;
  }>(url, { timeout: 300000 }); // 5 min — AWS pricing JSON is 300-400MB

  const specs: Record<string, InstanceSpecs> = {};
  for (const product of Object.values(resp.data.products)) {
    const attrs = product.attributes;
    if (
      attrs.operatingSystem === "Linux" &&
      attrs.tenancy === "Shared" &&
      attrs.capacitystatus === "Used" &&
      attrs.preInstalledSw === "NA" &&
      attrs.instanceType &&
      attrs.vcpu &&
      attrs.memory
    ) {
      const itype = attrs.instanceType;
      if (itype in specs) continue; // already recorded
      const vcpu = parseInt(attrs.vcpu, 10);
      // memory is like "8 GiB"
      const ramMatch = attrs.memory.match(/([\d.]+)\s*GiB/i);
      const ram_gb = ramMatch ? parseFloat(ramMatch[1]) : 0;
      if (!vcpu || !ram_gb) continue;
      const family = itype.split(".")[0];
      specs[itype] = { vcpu, ram_gb, family };
    }
  }
  return specs;
}

async function getRegionSpecs(region: string): Promise<Record<string, InstanceSpecs>> {
  // Return cached result immediately
  if (specsCache.has(region)) return specsCache.get(region)!;

  // If a fetch is already in-flight for this region, wait for it instead of
  // launching a second concurrent 350MB download (which causes stream aborts).
  if (inFlight.has(region)) return inFlight.get(region)!;

  const promise = fetchRegionSpecs(region)
    .then(specs => {
      specsCache.set(region, specs);
      inFlight.delete(region);
      return specs;
    })
    .catch(e => {
      console.error(`[aws-instances] Failed to fetch specs for ${region}:`, e);
      specsCache.set(region, {});
      inFlight.delete(region);
      return {} as Record<string, InstanceSpecs>;
    });

  inFlight.set(region, promise);
  return promise;
}

export async function getInstanceSpecs(
  instanceType: string,
  region = "us-east-1"
): Promise<InstanceSpecs | null> {
  const specs = await getRegionSpecs(region);
  return specs[instanceType] ?? null;
}

export async function getAllInstancesSortedByPrice(
  families: string[] | null,
  prices: Record<string, number>,
  region = "us-east-1"
): Promise<string[]> {
  const specs = await getRegionSpecs(region);
  const candidates = Object.keys(specs).filter(
    (itype) => families === null || families.includes(specs[itype].family)
  );
  return candidates.sort((a, b) => {
    const pa = prices[a] ?? Infinity;
    const pb = prices[b] ?? Infinity;
    return pa - pb;
  });
}

export async function suggestRightSizedInstance(
  cpuP95Pct: number,
  ramAvgPct: number,
  currentInstance: string,
  prices: Record<string, number>,
  region = "us-east-1"
): Promise<{ suggested: string; already_right_sized: boolean }> {
  const specs = await getRegionSpecs(region);
  const currentSpecs = specs[currentInstance];
  if (!currentSpecs) return { suggested: currentInstance, already_right_sized: true };

  const requiredVcpu = (cpuP95Pct / 100) * currentSpecs.vcpu * 1.3;
  const requiredRamGb = (ramAvgPct / 100) * currentSpecs.ram_gb * 1.3;

  const candidates = await getAllInstancesSortedByPrice(CANDIDATE_FAMILIES_V1, prices, region);

  for (const candidate of candidates) {
    const s = specs[candidate];
    if (!s) continue;
    if (s.vcpu >= requiredVcpu && s.ram_gb >= requiredRamGb) {
      if (candidate === currentInstance) return { suggested: currentInstance, already_right_sized: true };
      return { suggested: candidate, already_right_sized: false };
    }
  }
  return { suggested: currentInstance, already_right_sized: true };
}

export function computeEfficiencyScore(cpuAvg: number | null, ramAvg: number | null): number {
  if (cpuAvg === null || ramAvg === null) return 0;
  return Math.min(100, Math.max(0, Math.round(cpuAvg * 0.5 + ramAvg * 0.5)));
}

File: packages/backend/agents/tools/aws_instances.py
"""
EC2 instance catalog + right-sizing logic.
Reads from agents/config/ec2_instances.json (bundled, no network call).
"""
import json
import os
from pathlib import Path

_CATALOG_PATH = Path(__file__).parent.parent / "config" / "ec2_instances.json"
_catalog: dict | None = None

CANDIDATE_FAMILIES_V1 = ["t3", "t3a", "m5", "m5a", "c5", "r5"]


def _load_catalog() -> dict:
    global _catalog
    if _catalog is None:
        with open(_CATALOG_PATH) as f:
            _catalog = json.load(f)
    return _catalog


def get_instance_specs(instance_type: str) -> dict | None:
    """Return {vcpu, ram_gb, family} for an instance type, or None if not in catalog."""
    catalog = _load_catalog()
    return catalog.get(instance_type)


def get_all_instances_sorted_by_price(
    region: str,
    families: list[str] | None = None,
    prices: dict[str, float] | None = None,
) -> list[str]:
    """
    Return instance type names sorted by on-demand price ascending.
    `prices` is a dict of {instance_type: monthly_price} pre-fetched by the caller.
    If prices is None or an instance has no price, it is sorted to the end.
    Optionally filter by families list.
    """
    catalog = _load_catalog()
    candidates = [
        itype for itype, specs in catalog.items()
        if families is None or specs.get("family") in families
    ]
    def sort_key(itype):
        if prices and itype in prices:
            return prices[itype]
        return float("inf")
    return sorted(candidates, key=sort_key)


def suggest_right_sized_instance(
    cpu_p95_pct: float,
    ram_avg_pct: float,
    current_instance: str,
    prices: dict[str, float],
) -> dict:
    """
    Cross-family right-sizing across CANDIDATE_FAMILIES_V1.
    cpu_p95_pct: Datadog system-wide CPU % (0-100, no normalization needed)
    ram_avg_pct: avg RAM used as % of current instance's RAM GB
    current_instance: e.g. "r5.xlarge"
    prices: {instance_type: monthly_usd} for all candidates

    Returns:
      {"suggested": str, "already_right_sized": bool}
    """
    current_specs = get_instance_specs(current_instance)
    if not current_specs:
        return {"suggested": current_instance, "already_right_sized": True}

    required_vcpu = (cpu_p95_pct / 100) * current_specs["vcpu"] * 1.3
    required_ram_gb = (ram_avg_pct / 100) * current_specs["ram_gb"] * 1.3

    candidates = get_all_instances_sorted_by_price(
        region="us-east-1",
        families=CANDIDATE_FAMILIES_V1,
        prices=prices,
    )

    for candidate in candidates:
        specs = get_instance_specs(candidate)
        if not specs:
            continue
        if specs["vcpu"] >= required_vcpu and specs["ram_gb"] >= required_ram_gb:
            if candidate == current_instance:
                return {"suggested": current_instance, "already_right_sized": True}
            return {"suggested": candidate, "already_right_sized": False}

    # Fallback: no smaller candidate fits — already right-sized
    return {"suggested": current_instance, "already_right_sized": True}


def compute_efficiency_score(cpu_avg: float | None, ram_avg: float | None) -> int:
    """Weighted average utilization score 0-100. Returns 0 if either metric is None."""
    if cpu_avg is None or ram_avg is None:
        return 0
    raw = (cpu_avg * 0.5) + (ram_avg * 0.5)
    return int(min(100, max(0, raw)))


def efficiency_label(score: int, cpu_avg: float | None, ram_avg: float | None) -> str:
    if cpu_avg is None or ram_avg is None:
        return "unknown"
    if score < 30:
        return "over-provisioned"
    if score < 70:
        return "right-sized"
    return "under-provisioned"
